File: solutions/Merge-k-Sorted-Lists/Merge-k-Sorted-Lists.py


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param a list of ListNode
    # @return a ListNode
    def mergeKLists(self, lists):
        if lists==None or len(lists)==0:
            return None
        return self.mergelist(lists,0,len(lists)-1)
    def mergelist(self,lists,l,r):
        if l<r:
            m = (l+r)/2
            return self.merge(self.mergelist(lists,l,m),self.mergelist(lists,m+1,r))
        return lists[l]
    def merge(self,l1,l2):
        dummy = ListNode(0)
        dummy.next=l1
        cur = dummy
        while l1!=None and l2!=None:
            if l1.val<l2.val:
                l1=l1.next
            else:
                next=l2.next
                cur.next=l2
                l2.next=l1
                l2=next
            cur=cur.next
        if l2!=None:
            cur.next=l2
        return dummy.next
 

File: solutions/Merge-k-Sorted-Lists/Merge-k-Sorted-Lists.js
const merge = (l1, l2) => {
  const head = new ListNode(0);
  let p = head;
  while (l1 && l2) {
    if (l1.val < l2.val) {
      p.next = l1;
      p = l1;
      l1 = l1.next;
    } else {
      p.next = l2;
      p = l2;
      l2 = l2.next;
    }
  }
  p.next = l1 ? l1 : l2;
  return head.next;
};


const mergeKListsHelper = (lists, left, right) => {
  if (left < right) {
    const middle = left + Math.floor((right - left) / 2);
    return merge(mergeKListsHelper(lists, left, middle), mergeKListsHelper(lists, middle + 1, right));
  }
  return lists[left];
};

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function (lists) {
  if (lists.length === 0) {
    return null;
  }

  return mergeKListsHelper(lists, 0, lists.length - 1);
};